csi/moac/finalizer_helper.ts

```diff
@@ -0,0 +1,173 @@
+//
+'use strict';
+
+const k8s = require('@kubernetes/client-node');
+const log = require('./logger').Logger('finalizer_helper');
+
+export class FinalizerHelper {
+  private kubeConfig: any;
+  private k8sApi: any;
+  private namespace: String;
+  private groupname: String;
+  private version: String;
+  private plural: String;
+
+  constructor (namespace: String, groupname:String, version:String, plural:String) {
+      this.namespace = namespace;
+      this.groupname = groupname;
+      this.version = version;
+      this.kubeConfig = new k8s.KubeConfig();
+      this.kubeConfig.loadFromDefault();
+      this.k8sApi = this.kubeConfig.makeApiClient(k8s.CustomObjectsApi);
+      this.plural = plural;
+  }
+
+  addFinalizer(body: any, instancename: String, finalizer: String) {
+    if (body.metadata.deletionTimestamp != undefined) {
+      log.warn(`addFinalizer(${instancename},${finalizer}), deletionTimestamp is set`);
+      return;
+    }
+
+    if (body.metadata.finalizers != undefined) {
+      const index = body.metadata.finalizers.indexOf(finalizer);
+      if ( index > -1) {
+        log.debug(`@addFinalizer(${instancename},${finalizer}), finalizer already present`);
+        return;
+      }
+      body.metadata.finalizers.push(finalizer);
+    } else {
+      body.metadata.finalizers = [finalizer];
+    }
+
+    // TODO: use patchNamespacedCustomObject
+    this.k8sApi.replaceNamespacedCustomObject(
+        this.groupname,
+        this.version,
+        this.namespace,
+        this.plural,
+        instancename,
+        body)
+        .then((res:any) => {
+          log.debug(`added finalizer:${finalizer} to ${this.plural}:${instancename}`);
+        })
+        .catch((err:any) => {
+         log.error(`add finalizer:${finalizer} to ${this.plural}:${instancename}, update failed: code=${err.body.code}, reason=${err.body.reason}, ${err.body.message}`);
+        });
+  }
+
+  removeFinalizer(body: any, instancename: String, finalizer: String) {
+    if (body.metadata.finalizers == undefined) {
+        log.debug(`removeFinalizer(${instancename},${finalizer}), no finalizers on pool`);
```

Comment:
Should "pool" be specifically mentioned here? Everything else seems to be generic.